diff -ru ./automake-1.9.6.orig/lib/config.guess ./automake-1.9.6/lib/config.guess
--- ./automake-1.9.6.orig/lib/config.guess	2008-02-06 11:41:04 +0100
+++ ./automake-1.9.6/lib/config.guess	2008-02-06 11:52:26 +0100
@@ -779,8 +779,14 @@
     i*:PW*:*)
 	echo ${UNAME_MACHINE}-pc-pw32
 	exit ;;
-    x86:Interix*:[34]*)
-	echo i586-pc-interix${UNAME_RELEASE}|sed -e 's/\..*//'
+    x86:Interix*:[3456]*)
+       eval $set_cc_for_build
+       echo "__PARITY__" > $dummy.c
+       if $CC_FOR_BUILD -E $dummy.c | grep "__PARITY__" > /dev/null 2>&1; then
+           echo i586-pc-interix${UNAME_RELEASE}
+       else
+           echo i586-pc-winnt
+       fi
 	exit ;;
     [345]86:Windows_95:* | [345]86:Windows_98:* | [345]86:Windows_NT:*)
 	echo i${UNAME_MACHINE}-pc-mks
