diff -ru ./automake-1.10.orig/lib/config.guess ./automake-1.10/lib/config.guess
--- ./automake-1.10.orig/lib/config.guess	2008-02-06 11:53:40 +0100
+++ ./automake-1.10/lib/config.guess	2008-02-06 11:54:02 +0100
@@ -791,7 +791,13 @@
 	echo ${UNAME_MACHINE}-pc-pw32
 	exit ;;
     x86:Interix*:[3456]*)
+	   eval $set_cc_for_build
+	   echo "__PARITY__" > $dummy.c
+	   if $CC_FOR_BUILD -E $dummy.c | grep "__PARITY__" > /dev/null 2>&1; then
 	echo i586-pc-interix${UNAME_RELEASE}
+	   else
+		   echo i586-pc-winnt
+	   fi
 	exit ;;
     EM64T:Interix*:[3456]*)
 	echo x86_64-unknown-interix${UNAME_RELEASE}
