diff -ru gnuconfig.orig/config.guess gnuconfig/config.guess
--- gnuconfig.orig/config.guess	2008-02-06 11:57:30 +0100
+++ gnuconfig/config.guess	2008-02-06 11:59:05 +0100
@@ -792,7 +792,13 @@
 	echo ${UNAME_MACHINE}-pc-pw32
 	exit ;;
     x86:Interix*:[3456]*)
-	echo i586-pc-interix${UNAME_RELEASE}
+       eval $set_cc_for_build
+       echo "__PARITY__" > $dummy.c
+       if $CC_FOR_BUILD -E $dummy.c | grep "__PARITY__" > /dev/null 2>&1; then
+	       echo i586-pc-interix${UNAME_RELEASE}
+       else
+           echo i586-pc-winnt
+       fi
 	exit ;;
     EM64T:Interix*:[3456]* | authenticamd:Interix*:[3456]*)
 	echo x86_64-unknown-interix${UNAME_RELEASE}
